apps/prairielearn/src/lib/exam-mode.ts

```diff
@@ -46,10 +46,14 @@ export async function ipToMode({
   date,
   authn_user_id,
 }: {
-  ip: string;
+  ip: string | undefined;
   date: Date;
   authn_user_id: string;
 }) {
+  // Express's types indicate that `ip` may be undefined in some cases. We want
+  // to ensure that we don't try to proceed without one.
+  if (ip === undefined) throw new Error('IP address is required');
```

Comment:
Alternatively: we could permit it to go to the database, in which case it'd be `NULL`, and that should also deny access. I'm not sure how common missing IPs are in practice (probably not very).